### PR TITLE
[JENKINS-68286] Allow extra classes with <layout:icon>

### PR DESCRIPTION
**Pull request by ikedam**
See [JENKINS-68286](https://issues.jenkins.io/browse/JENKINS-68286).

`<layout:icon>`  decides icons to display with `class` attribute.
It fails to decide icon if you specify extra classes. But extra classes are often useful apply css.

This change does followings:

* Suppose icon classes start with `icon-`.
* Remove classes without `icon-` prefixes and retry decide icons if fail to decide icons.

### Proposed changelog entries

* Allow extra CSS classes with `<l:icon>`.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change) and are in the imperative mood. [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadoc, as appropriate. : N/A
- [x] For dependency updates: links to external changelogs and, if possible, full diffs: N/A

### Desired reviewers

N/A

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are accurate, human-readable, and in the imperative mood
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

### CODE REVIEW

File: core/src/main/java/hudson/Functions.java
Comment (ikedam) [inline core/src/main/java/hudson/Functions.java:2293]: I don't know what this code does.
But I think this code isn't necessary and removed it to make it easy to run tests.

File: core/src/main/java/hudson/Functions.java
Comment (ikedam) [inline core/src/main/java/hudson/Functions.java:2303]: `icon-settings` is unavailable.

File: core/src/test/java/hudson/FunctionsTest.java
Comment (ikedam) [inline core/src/test/java/hudson/FunctionsTest.java:663]: This will fail without the change of `tryGetIcon`:

```
[ERROR] tryGetIcon_shouldReturnMetadataForExtraSpec  Time elapsed: 0.018 s  <<< FAILURE!
java.lang.AssertionError:

Expected: is not null
     but: was null
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:6)
        at hudson.FunctionsTest.tryGetIcon_shouldReturnMetadataForExtraSpec(FunctionsTest.java:662)
```

File: core/src/main/java/hudson/Functions.java
Comment (timja) [inline core/src/main/java/hudson/Functions.java:2297]:
```suggestion
        if (iconMetadata == null && iconGuess.contains(" ")) {
```

**Review by timja — COMMENTED** [review]
FWIW we've been working around this by wrapping elements in `spans` and applying the css to the wrapped element.

I need to check how this behaves with translated symbols before approving

**Review by NotMyFault — COMMENTED** [review]
> FWIW we've been working around this by wrapping elements in spans and applying the css to the wrapped element.

That would be something to consider nevertheless to reintroduce an easy way to apply styles on icons.
Based on a quick test, icon -> symbol translations are working fine still. This is part of the workflow-support-plugin:
Bildschirmfoto 2022-05-03 um 10 46 08

**Review by timja — APPROVED** [review]

**Review by NotMyFault — APPROVED** [review]

**Comment by timja** [comment]
This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Pull request merged**
